include/revng/Model/Type/BaseType.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "revng/Model/GeneratedType.h"
+
+/// \brief Concept to identify all types that are derived from model::Type
+template<typename T>
+concept IsModelType = std::is_base_of_v<model::Type, T>;
+
+/// \brief Base class of model types used for LLVM-style RTTI
+class model::Type : public model::generated::Type {
+public:
+  // TODO: Constructors cannot be inherited, since the default one is
+  //  manually implemented in order to generate a random ID
+  Type();
+  Type(TypeKind::Values TK);
+  Type(TypeKind::Values Kind, uint64_t ID) : model::generated::Type(Kind, ID) {}
+
+public:
+  static bool classof(const Type *T) { return false; }
+
+  Key key() const { return Key{ Kind, ID }; }
+
+  Identifier name() const;
+
+public:
+  bool operator==(const Type &Other) const { return key() == Other.key(); }
+
+  bool operator<(const Type &Other) const { return key() < Other.key(); }
+
+public:
+  std::optional<uint64_t> size() const debug_function;
+  RecursiveCoroutine<std::optional<uint64_t>> size(VerifyHelper &VH) const;
+
+public:
+  bool verify() const debug_function;
+  bool verify(bool Assert) const debug_function;
+  RecursiveCoroutine<bool> verify(VerifyHelper &VH) const;
+  void dump() const debug_function;
+
+public:
+  // TODO: we should not need to define this here
+  static constexpr const char *Tag = "!Type";
+  static constexpr const auto AssociatedKind = TypeKind::Invalid;
```

Comment:
One day we might want to generate `classof` and `AssociatedKind`. But at this stage, I'm not sure it's easy due to `FunctionEdge`/`CallEdge`.